source/mir/sparse/blas/dot.d

```diff
@@ -0,0 +1,173 @@
+module mir.sparse.blas.dot;
+
+import std.traits;
+import mir.ndslice.slice;
+import mir.sparse;
+
+
+/++
+Params:
+	x = sparse vector
+	y = sparse vector
+Returns:
+	scalar `x × y`
++/
+Unqual!(CommonType!(T1, T2)) dot(
+	V1 : CompressedArray!(T1, I1),
+	V2 : CompressedArray!(T2, I2),
+	T1, T2, I1, I2)
+(V1 x, V2 y)
+{
+	return dot!(typeof(return))(x, y);
+}
+
+/// ditto
+D dot(
+	D,
+	V1 : CompressedArray!(T1, I1),
+	V2 : CompressedArray!(T2, I2),
+	T1, T2, I1, I2)
+(V1 x, V2 y)
+{
+	pragma(inline, false);
+
+	typeof(return) s = 0;
+
+	uint done = 2;
+	Unqual!I1 ai0 = void;
+	Unqual!I2 bi0 = void;
+
+	if(x.indexes.length && y.indexes.length) for (;;)
+	{
+		bi0 = y.indexes[0];
+		if (x.indexes[0] < bi0)
+		{
+			do
+			{
+				x.values = x.values[1 .. $];
+				x.indexes = x.indexes[1 .. $];
+				if (x.indexes.length == 0)
+				{
+					break;
+				}
+			}
+			while(x.indexes[0] < bi0);
+			done = 2;
+		}
+		if (--done == 0)
+		{
+			goto L;
+		}
+		ai0 = x.indexes[0];
+		if (y.indexes[0] < ai0)
+		{
+			do
+			{
+				y.values = y.values[1 .. $];
+				y.indexes = y.indexes[1 .. $];
+				if (y.indexes.length == 0)
+				{
+					break;
+				}
+			}
+			while(y.indexes[0] < ai0);
+			done = 2;
+		}
+		if (--done == 0)
+		{
+			goto L;
+		}
+		continue;
+		L:
+		s += D(x.values[0]) * D(y.values[0]);
+		x.indexes = x.indexes[1 .. $];
+		if(x.indexes.length == 0)
+		{
+			break;
+		}
+		y.indexes = y.indexes[1 .. $];
+		if(y.indexes.length == 0)
+		{
+			break;
+		}
+		x.values = x.values[1 .. $];
+		y.values = y.values[1 .. $];
+	}
+
+	return s;
+}
+
+///
+unittest
+{
+	auto x = CompressedArray!(int, uint)([1, 3, 4, 9, 10], [0, 3, 5, 9, 100]);
+	auto y = CompressedArray!(int, uint)([1, 10, 100, 1000], [1, 3, 4, 9]);
+	assert(dot(x, y) == 9030);
```

Comment:
maybe you want to add the vector representation in the doc - it's a lot easier to read ;-)
```
x = [1,0,0, 3,0,4,0,0,0,9,.....,10]
y = [0,1,0,10,0,0,0,0,0,1000]
```
= 30 * 9000